### ISSUE TITLE
Mouse not trapped in DDNet window

### ISSUE THREAD

**Issue by phcrm** [ddnet/ddnet#2091]
**Problem**: When focusing another application and later return to DDNet the mouse sometimes isn't trapped inside client window. Getting another courser which moves outside the window to next monitor and focuses window there.
Sadly can't reliably reproduce the issue as it doesn't always happen.

**Current Solution**: Only fix currently is to restart the client.

**More Infos**: 
- Am using [bspwm](https://github.com/baskerville/bspwm) which supports subset of EWMH and ICCCM.
- Client version is `12.9.2`

**Comment by def-** [comment]
I've noticed the same thing a few times with xmonad. I thought it was a problem of my WM, but apparently not. My guess would be an SDL regression since I don't think we changed anything in that code, and I didn't have this a few years ago.

**Comment by Jupeyy** [comment]
Could be unrelated:
When I tested to move everything in a new thread (except SDL), i noticed this too sometimes.
Maybe the clients notifys SDL too early(/late or whatever) about the mouse state.

In my case a simple F1 to toggle the console(which uses another mouse state) fixed it.

**Comment by def-** [comment]
nouaa on discord reports the same.

https://user-images.githubusercontent.com/2335377/150657005-3148b633-fb1a-4981-9859-289b5c71788b.mp4

**Comment by Jupeyy** [comment]
If it's at startup it might be fixed with SDL 2.0.18
`https://github.com/libsdl-org/SDL/pull/4293#issuecomment-934840693`

**Comment by heinrich5991** [comment]
@Speedy-Consoles told me that he only experiences it with the Steam client, not with the website one. Can you confirm @Speedy-Consoles?

**Comment by Speedy-Consoles** [comment]
I didn't try the website client, but one built from source (some revision newer than the Steam client). It didn't have the issue.
To me it looked like it depends on the amount of time the client isn't in focus. If it's only for like 5 seconds, the cursor stays captured. If it gets too long, it escapes.
Only a restart of the client fixes it for me. F1 doesn't help.

**Comment by NouaaTW** [comment]
on my side, when this "bug" appeared I was with the steam version

**Comment by Jupeyy** [comment]
> on my side, when this "bug" appeared I was with the steam version

Do you remember if it only happens after you cleanly started the client, so never tabbed out and in again?

I remember that steam opens a small window "starting game" that could change the focus in last second before client starts and caused problems (on my emulator).
And i think the SDL pr i posted wanted to fix that

**Comment by NouaaTW** [comment]
@Jupeyy yes, i think i tabbed out when the game was starting, but im not sure

**Comment by SollyBunny** [comment]
I've never had this on DWM, I think this should be a feature when you're not in game or menu isn't opened

**Comment by ChillerDragon** [comment]
I can reproduce this consistently by opening two clients and switching between them. The clients are self compiled at current ddnet master branch. Using `gfx_fullscreen 3` and a outdated archlinux system with gnome.

**Comment by AssassinTee** [comment]
I can reproduce this but not very consistently only in the steam client and with 2 monitors (the mouse always ends up on the second monitor). If I build the client myself I don't have this issue, so this might also be related to steam and their runtime or **maybe** proton (which I doubt because we are linux native)

**Comment by solenum** [comment]
> I can reproduce this but not very consistently only in the steam client and with 2 monitors (the mouse always ends up on the second monitor). If I build the client myself I don't have this issue, so this might also be related to steam and their runtime or **maybe** proton (which I doubt because we are linux native)

It's a SDL3 bug, I can reproduce when built against sdl2-compat, which is a compatability layer that actually uses sdl3 under the hood, when I build the same thing against sdl2 directly the bug doesn't exist, I did my best to trace the point where this issue starts by running through with a debugger and I could see it happening somewhere in a SDL call, nothing to do with the ddnet code or steam runtime (nor proton, which isn't used here as this is a native application).

If you scan the sdl3 issue tracker you'll find a few related issues, I've not found a concrete workaround yet aside from ensuring I build against sdl2 and not sdl2-compat/sdl3

**Comment by AssassinTee** [comment]
> built against sdl2-compat

My self compiled client also builds against sdl2-compat, maybe it's a version difference

**Comment by solenum** [comment]
Yes this issue is fairly old, I haven't tested against the latest sdl3 so it may have been solved by now.  Might be wise to try and compile against a sdl3/sdl2-compat version from around the time of my original comment and see if the issue is present, and then bisect the versions and re-build until you find roughly where it was solved to understand the cause.
